### ISSUE TITLE
Plan: Improve Handling of Sparse 2μm Data & Masking in Multiscale Training

### ISSUE THREAD

**Issue by vanbelkummax** [vanbelkummax/visium-hd-2um-benchmark#1]
## Overview
This issue documents a prioritized plan to address gaps in the handling of masked (empty) bins, supervision, and evaluation for 2μm spatial transcriptomics data within the multiscale training pipeline.  The plan is based on a review of current code and recent suggestions, and is meant to increase model robustness—especially in avoiding over-prediction in empty regions, improving 2μm-specific evaluation, and ensuring consistent comparison across code paths.

---

## High-Impact Next Steps (Top Priority)

1. **Explicit Empty-Bin Penalty**
   - Add a penalty so masked-out (empty) 2μm bins are nudged toward zero, e.g.:   
     ```python
     inv_mask = ~mask_2um
     inv_mask_exp = inv_mask.unsqueeze(1).expand_as(pred)
     loss_empty = F.mse_loss(pred[inv_mask_exp], torch.zeros_like(pred[inv_mask_exp]))
     total = loss_8um + lambda_2um * loss_2um + lambda_empty * loss_empty
     ```
   - Rationale:  Prevents model from over-predicting signal in bins that genuinely lack data.

2. **Explicit 2μm Metrics and False Positive Rate**
   - Add metrics for 2μm (masked PCC, SSIM) and a "false-positive in empty bins" metric during evaluation.
   - Rationale: Current LOOCV metrics only evaluate 8μm upsampling; over-prediction in empty bins can go undetected.

3. **Align Decoder Comparison Scripts to Use Per-Bin Masking**
   - `train_decoder_comparison_128.py` and `quick_decoder_compare. py` should use `patches_multiscale. npy` and per-bin `mask_128` rather than patch-level `labels_2um` for fair, sparse supervision.

4. **Weight 2μm Loss by Coverage**
   - Use `coverage_2um` from multiscale patches to scale `lambda_2um` for each patch by `mask. sum()/mask.numel()`.
   - Rationale: Prevents low-coverage patches from dominating training gradients.

---

## Medium-Impact Next Steps

5. **Anneal lambda_2um**
   - Start with `lambda_2um≈0` and ramp up once the 8μm loss stabilizes.
   - Rationale: This improves stability with very sparse supervision.

6. **Try Robust Loss for Sparse Targets**
   - Use `HuberLoss`, `Charbonnier`, or log1p transform for 2μm to reduce dominance of near-zero bins.

7. **Alternative: L1 Sparsity Regularizer**
   - If not using a full occupancy head, add L1 regularization on the model's output in masked-out regions as a lighter alternative to the empty-bin MSE penalty.

---

## Potential Additions

- **Occupancy Head (Binary Data-Presence Classifier):**
   - Add a binary "data present" head, supervised by `mask_2um`, and gate 2μm expression predictions by this occupancy signal.
   - Candidate location: model/img2st_enhanced.py (add occupancy head, BCEWithLogitsLoss, and gating logic).

---

## Clarification Needed

- **Definition of mask_128:**
   - Current code uses mask_128 as a binary "barcode/data present" indicator; thus, penalizing predictions in masked-out bins is correct and preferred.

---

## Action
- Begin with high-impact items 1–4, using new 2μm metrics to validate effectiveness.
- Address medium-impact and potential additions as needed based on evaluation results.

---

_This plan is based on code review as of Dec 2025. Please update this issue as work progresses._
